src/gc-interface-allocation.h

```diff
@@ -0,0 +1,182 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+#ifndef JL_GC_INTERFACE_ALLOCATION_H
+#define JL_GC_INTERFACE_ALLOCATION_H
+
+#include <stdint.h>
+
+#include "julia_threads.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define JL_SMALL_BYTE_ALIGNMENT 16
+#define JL_CACHE_BYTE_ALIGNMENT 64
+// JL_HEAP_ALIGNMENT is the maximum alignment that the GC can provide
+#define JL_HEAP_ALIGNMENT JL_SMALL_BYTE_ALIGNMENT
+#define GC_MAX_SZCLASS (2032-sizeof(void*))
+
+struct _jl_value_t *jl_gc_pool_alloc_noinline(jl_ptls_t ptls, int pool_offset, int osize);
+struct _jl_value_t *jl_gc_big_alloc_noinline(jl_ptls_t ptls, size_t allocsz);
+
+#ifndef THIRD_PARTY_GC
```

Comment:
It seems like this file isn't really an _interface_ file - it's an implementation, right? It's implementing the interface `jl_gc_alloc`, right?

I would maybe reorganize this so that this file simply contains the *declarations* for the allocation interface, and then if not THIRD_PARTY_GC, you include the native GC header.
Something like this:
```c
JL_DLLEXPORT jl_value_t *jl_gc_alloc(jl_ptls_t ptls, size_t sz, void *ty);
uint8_t JL_CONST_FUNC jl_gc_szclass(unsigned sz) JL_NOTSAFEPOINT;
...
```
then, either in this header, or maybe even better in julia-internal.h, you'd have:
```c
#ifndef THIRD_PARTY_GC
#include "gc-native-impl-allocation.h"
#endif
```
or something like that?
Basically, following something like what is suggested here: https://stackoverflow.com/a/9317494/751061

Reply:
Even better than that IMO would be to make the file an actual interface (e.g. implementation free), as you are suggesting, and move the implementation into a separate `.c` file.

This would be the best choice for keeping MMTk or any external backend as a separate library that we simply link into Julia, as mentioned above.

IIRC, inning across different `.o` files is not possible unless you have LTO though...

Reply:
Yeah. I think it's fine to have this be a build-time choice, not a link-time choice. I agree that would be the coolest, if you didn't have to rebuild julia, but the reality is that for inlining perf we probably will have to make it a build-time choice, yeah 👍 

You should check in with the ANU folks on these details; i think they have a lot of experience with the various options available here 😊